src/join/joining.cu

```diff
@@ -307,16 +309,199 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
+        default:            return 0;
+    }
 }
 
-gdf_error gdf_inner_join(int num_cols, gdf_column **leftcol, gdf_column **rightcol,
-                         gdf_column *left_result, gdf_column *right_result,
-                         gdf_context *join_context)
-{
-  return join_call<JoinType::INNER_JOIN>(num_cols, leftcol, rightcol, left_result, right_result, join_context);
+template <JoinType join_type, typename size_type>
+gdf_error join_call_compute_df(
+                         gdf_column **left_cols, 
+                         int num_left_cols,
+                         int left_join_cols[],
+                         gdf_column **right_cols,
+                         int num_right_cols,
+                         int right_join_cols[],
+                         int num_cols_to_join,
+                         int result_num_cols,
+                         gdf_column **result_cols,
+                         gdf_column * left_indices,
+                         gdf_column * right_indices,
+                         gdf_context *join_context) {
+    //return error if the inputs are invalid
+    if ((left_cols == nullptr)  ||
+        (right_cols == nullptr)) { return GDF_DATASET_EMPTY; }
+
+    //check if combined join output is expected
+    bool compute_df = (result_cols != nullptr);
+
+    //return error if no output pointers are valid
+    if ( ((left_indices == nullptr)||(right_indices == nullptr)) &&
+         (!compute_df) ) { return GDF_DATASET_EMPTY; }
+
+    //If index outputs are not requested, create columns to store them
+    //for computing combined join output
+    gdf_column * left_index_out = left_indices;
+    gdf_column * right_index_out = right_indices;
+    std::vector<std::unique_ptr<gdf_column>> temporary_gdf_columns;
+    if (left_indices == nullptr) {
+        std::unique_ptr<gdf_column> ptr(new gdf_column);
+        temporary_gdf_columns.push_back(std::move(ptr));
+        left_index_out = temporary_gdf_columns.back().get();
+    }
+    if (right_indices == nullptr) {
+        std::unique_ptr<gdf_column> ptr(new gdf_column);
+        temporary_gdf_columns.push_back(std::move(ptr));
+        right_index_out = temporary_gdf_columns.back().get();
+    }
+
+    //get column pointers to join on
+    std::vector<gdf_column*> ljoincol;
+    std::vector<gdf_column*> rjoincol;
+    for (int i = 0; i < num_cols_to_join; ++i) {
+        ljoincol.push_back(left_cols[ left_join_cols[i] ]);
+        rjoincol.push_back(right_cols[ right_join_cols[i] ]);
+    }
+    gdf_error join_err = join_call<join_type>(num_cols_to_join,
+            ljoincol.data(), rjoincol.data(),
+            left_index_out, right_index_out,
+            join_context);
+    //If compute_df is false then left_index_out or right_index_out
+    //was not dynamically allocated.
+    if ((!compute_df) || (GDF_SUCCESS != join_err)) {
+        return join_err;
+    }
+
+    //create left and right input table with columns not joined on
```

Comment:
Can you wrap the below functionality in a `construct_output_dataframe` function?